### DISCUSSION TITLE
How to remove CRLF characters when using grep_live?

### DISCUSSION THREAD

**Discussion by AmerM137** [nvim-mini/mini.nvim#1404] [Q&A] [answered]
I am seeing "^M" characters at the end of the lines when live grepping. I am on windows and these files do in fact have CRLF line endings, but I noticed that telescope removes them somehow.

How can we remove CRLF characters so they don't show when live grepping?

Thanks!

**Comment by echasnovski** [comment]
I am afraid, I don't have Windows, so we'd have to do a little pair investigation, if you don't mind. Here are initial questions:

- Would you mind comparing how executing `rg` looks in 'mini.pick' and manually in terminal emulator? Here is the ideal info:
    - Start `MiniPick.builtin.live_grep`, type some `pattern`, and make a screenshot.
    - Open terminal emulator (in the same current working directory) and execute `rg --column --line-number --no-heading --field-match-separator \\x00 --no-follow --color=never -- pattern` (replace `pattern` with pattern you used in previous step) and make a screenshot. Order of items might change, but that's ok. I am curious whether there are extra `^M` at the end there.
- What is the output of the following command (that prints values of certain options): `:echo &fileformat &fixendofline` ?

My guess (and hope) that this can be fixed with some `rg` flag, like `--crlf`. But not sure.

**Reply by AmerM137** [reply]
Here are the tests!

live_grep with mini.pick from within nvim:
<img width="511" alt="live_grep_crlf" src="https://github.com/user-attachments/assets/c38f33b3-f55f-49dd-b425-c7309602b078" />

rg command in terminal emulator
<img width="950" alt="rg_in_terminal" src="https://github.com/user-attachments/assets/676911ae-5c5b-40a2-994e-f97854a0d10b" />

output of `:echo &fileformat &fixendofline` is `dos 1`

**Reply by echasnovski** [reply]
Thanks for the follow up!

Unfortunately, it denies some early hope that it is just how 'ripgrep' displays things (then text would be visibly the same in 'mini.pick' and terminal) and that some relevant options are misconfigured (seems like they are as needed).

------

Would you mind trying to use `live_grep` with `--crlf` flag? For the quick test, append [this line](https://github.com/echasnovski/mini.nvim/blob/7ebfab26d77a4b9b05aaae565907e7fa4b2ee154/lua/mini/pick.lua#L3291) with the `, '--crlf'` text. Do that in 'lua/mini/pick.lua' file inside the installed 'mini.nvim' (or 'mini.pick') plugin that is loaded in your session (depends on how you installed it).

**Comment by AmerM137** [comment]
Just tried it. Unfortunately, it doesn't seem to work. My understanding is that the "--crlf" flag adds crlf line terminators. Do we need to post-process the results somehow to convert the crlf to lf endings? Maybe that's what telescope does.

Here is what I changed, I am using mini.deps so I found pick.lua in my pack/deps/start...
<img width="772" alt="image" src="https://github.com/user-attachments/assets/2dd13d00-79c6-4422-9f56-eaf00fefbb48" />

**Reply by AmerM137** [reply]
fzf-lua seems to strip the crlf endings just fine @echasnovski 
<img width="1340" alt="image" src="https://github.com/user-attachments/assets/7cbb330c-7bce-45cc-b9d3-2116dbc9e913" />

**Reply by echasnovski** [reply, marked as answer]
> fzf-lua seems to strip the crlf endings just fine

Yes, because it seems to [adjust that during file reading regardless of platform](https://github.com/ibhagwan/fzf-lua/blob/8e6a1e9f92b531ee965db4713c0227aeb37b4a22/lua/fzf-lua/previewer/builtin.lua#L841). I think I'd like a more OS robust approach if it is possible.

> Do we need to post-process the results somehow to convert the crlf to lf endings?

Yes, adjusting how items are shown was next step. Would you mind, please, checking two possible solutions to this? Apply each of these independently from one another:
- Replace [this line](https://github.com/echasnovski/mini.nvim/blob/7ebfab26d77a4b9b05aaae565907e7fa4b2ee154/lua/mini/pick.lua#L1711) with `    local items = vim.split(table.concat(data_feed), H.is_windows and '\r\n' or '\n')`. This one adjusts how lines are separated when getting results from CLI.
- Replace [this line](https://github.com/echasnovski/mini.nvim/blob/7ebfab26d77a4b9b05aaae565907e7fa4b2ee154/lua/mini/pick.lua#L1064) with `  lines = vim.tbl_map(function(l) return l:gsub('%z', '│'):gsub('\r$', ''):gsub('[\n\r]', ' '):gsub('\t', tab_spaces) end, lines)`. This one adjusts how items are shown.

My hope is that both work, but you never know.

**Reply by AmerM137** [reply]
Ok tested each one independently, results:
Replacing 1: file picker gets messed up, shows all results on one line, but CRLF are now stripped in live grep.
Replacing 2: works! file picker is still good, CRLF are stripped in live grep.

2 is the winner.

**Reply by AmerM137** [reply]
I tried the first option again with `local items = vim.split(table.concat(data_feed), '[\r]?\n')` (a la fzf-lua) and it works as well. Let me know what else you'd like me to try to make option 1 work and keep it OS robust.

**Comment by echasnovski** [comment]
This should now be fixed on latest `main`. I adopted both from [this comment](https://github.com/echasnovski/mini.nvim/discussions/1404#discussioncomment-11621498) with slight modifications based on your feedback. Thanks for all the feedback, it really helped!
